lapce-data/src/editor/commands/insert_chars.rs

```diff
@@ -0,0 +1,291 @@
+                if cursor_char == Some(c) {
+                    // Skip the closing character
+                    let new_offset = buffer.buffer.next_grapheme_offset(
+                        offset,
+                        1,
+                        buffer.buffer.len(),
+                    );
+
+                    *region = SelRegion::caret(new_offset);
+                    continue;
+                }
+
+                let line = buffer.buffer.line_of_offset(offset);
+                let line_start = buffer.buffer.offset_of_line(line);
+                if buffer.buffer.slice_to_cow(line_start..offset).trim() == "" {
+                    let opening_character = matching_char(c).unwrap();
+                    if let Some(previous_offset) = buffer.buffer.previous_unmatched(
+                        syntax.as_ref(),
+                        opening_character,
+                        offset,
+                    ) {
+                        // Auto-indent closing character to the same level as the opening.
+                        let previous_line =
+                            buffer.buffer.line_of_offset(previous_offset);
+                        let line_indent =
+                            buffer.buffer.indent_on_line(previous_line);
+
+                        let current_selection = region_to_selection(SelRegion::new(
+                            line_start, offset, None,
+                        ));
+
+                        edits.push((current_selection, format!("{line_indent}{c}")));
+                        continue;
+                    }
+                }
+            }
+
+            if matching_pair_type == Some(true) {
+                // Create a late edit to insert the closing pair, if allowed.
+                let is_whitespace_or_punct = cursor_char
+                    .map(|c| {
+                        let prop = get_word_property(c);
+                        prop == WordProperty::Lf
+                            || prop == WordProperty::Space
+                            || prop == WordProperty::Punctuation
+                    })
+                    .unwrap_or(true);
+
+                if is_whitespace_or_punct {
+                    let insert_after = matching_char(c).unwrap();
+                    edits_after.push((idx, insert_after));
+                }
+            };
+
+            let current_selection = region_to_selection(*region);
+
+            edits.push((current_selection, c.to_string()));
+        }
+
+        // Apply edits to current selection
+        let edits = edits
+            .iter()
+            .map(|(selection, content)| (selection, content.as_str()))
+            .collect::<Vec<_>>();
+
+        let delta = buffer.edit_multiple(&edits, EditType::InsertChars);
+
+        // Update selection
+        let mut selection =
+            selection.apply_delta(&delta, true, InsertDrift::Default);
+
+        // Apply late edits
+        let edits_after = edits_after
```

Comment:
Note to self: It would be better to allow edits of `SelRegion`s, not just `Selection`s. That way, we wouldn't have to create two vectors.